controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,221 @@ package controllers
+		rc.SetStatusCondition(createStatusConditionErrored(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "StardogInstance not ready"))
+		return ctrl.Result{Requeue: true}, r.updateStatus(sir)
+	}
+	rc.SetStatusCondition(createStatusConditionReady(true, "Synchronized"))
+	return ctrl.Result{Requeue: false}, r.updateStatus(sir)
+}
+
+func (r *StardogInstanceReconciler) deleteStardogInstance(sir *StardogInstanceReconciliation) error {
+	stardogInstance := sir.resource
+	rc := sir.reconciliationContext
+	credentials := sir.resource.Spec.AdminCredentials
+
+	r.Log.V(1).Info("retrieving admin credentials from Secret", "secret", credentials.Namespace+"/"+credentials.SecretRef)
+	username, password, err := rc.getCredentials(r.Client, credentials, rc.namespace)
+	if err != nil {
+		return err
+	}
+	stardogClient := rc.stardogClient
+	stardogClient.SetConnection(sir.resource.Spec.ServerUrl, username, password)
+
+	if contains(stardogInstance.GetFinalizers(), instanceUserFinalizer) {
+		if err := r.userFinalizer(stardogClient, rc.context); err != nil {
+			return err
+		}
+		controllerutil.RemoveFinalizer(stardogInstance, instanceUserFinalizer)
+	}
+	if contains(stardogInstance.GetFinalizers(), instanceRoleFinalizer) {
+		if err := r.roleFinalizer(stardogClient, rc.context); err != nil {
+			return err
+		}
+		controllerutil.RemoveFinalizer(stardogInstance, instanceRoleFinalizer)
+	}
+	if err := r.Update(rc.context, stardogInstance); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) userFinalizer(stardogClient stardogrestapi.ExtendedBaseClientAPI, ctx context.Context) error {
+	users, err := stardogClient.ListUsers(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot list Stardog users: %v", err)
+	}
+	if len(*users.Users) > 0 {
+		return fmt.Errorf("cannot delete StardogInstance, found %d users", len(*users.Users))
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) roleFinalizer(stardogClient stardogrestapi.ExtendedBaseClientAPI, ctx context.Context) error {
+	roles, err := stardogClient.ListRoles(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot list Stardog roles: %v", err)
+	}
+	if len(*roles.Roles) > 0 {
+		return fmt.Errorf("cannot delete StardogInstance, found %d roles", len(*roles.Roles))
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) validateSpecification(spec StardogInstanceSpec) error {
+	r.Log.V(1).Info("validating StardogInstanceSpec")
+	if spec.ServerUrl == "" {
+		return fmt.Errorf(".spec.ServerUrl is required")
+	}
+	if spec.AdminCredentials.SecretRef == "" {
+		return fmt.Errorf(".spec.AdminCredentials.SecretRef is required")
+	}
+	_, err := url.ParseRequestURI(spec.ServerUrl)
+	if err != nil {
+		return fmt.Errorf(".spec.ServerUrl is not a valid URL: %v", err)
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) validateConnection(sir *StardogInstanceReconciliation) error {
+	r.Log.Info("verifying connection to Stardog API")
```

Comment:
If you're going to log it, I would expect at least the actual URL also in the log. If the controller is managing multiple stardog instances, this becomes hard to distinguish from eachother